tensorpack/dataflow/serialize.py

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*-
+# File: serialize.py
+
+import os
+import numpy as np
+
+from ..utils.utils import get_tqdm
+from ..utils import logger
+from ..utils.serialize import dumps, loads
+
+from .base import DataFlow
+from .format import LMDBData, TFRecordData
+from .common import MapData
+from .raw import DataFromList
+
+__all__ = ['LMDBSerializer', 'NumpySerializer', 'TFRecordSerializer']
+
+
+def _reset_df_and_get_size(df):
+    df.reset_state()
+    try:
+        sz = df.size()
+    except NotImplementedError:
+        sz = 0
+    return sz
+
+
+class LMDBSerializer():
+    @staticmethod
+    def save(df, path, write_frequency=5000):
+        """
+        Dump a Dataflow to a lmdb database, where the keys are indices and values
+        are serialized datapoints.
+        The output database can be read directly by
+        :class:`tensorpack.dataflow.LMDBDataPoint`.
+
+        Args:
+            df (DataFlow): the DataFlow to dump.
+            path (str): output path. Either a directory or a mdb file.
+            write_frequency (int): the frequency to write back data to disk.
+        """
+        assert isinstance(df, DataFlow), type(df)
+        isdir = os.path.isdir(path)
+        if isdir:
+            assert not os.path.isfile(os.path.join(path, 'data.mdb')), "LMDB file exists!"
+        else:
+            assert not os.path.isfile(path), "LMDB file exists!"
+        db = lmdb.open(path, subdir=isdir,
+                       map_size=1099511627776 * 2, readonly=False,
+                       meminit=False, map_async=True)    # need sync() at the end
+        size = _reset_df_and_get_size(df)
+        with get_tqdm(total=size) as pbar:
+            idx = -1
+
+            # LMDB transaction is not exception-safe!
+            # although it has a context manager interface
+            txn = db.begin(write=True)
+            for idx, dp in enumerate(df.get_data()):
+                txn.put(u'{}'.format(idx).encode('ascii'), dumps(dp))
+                pbar.update()
+                if (idx + 1) % write_frequency == 0:
+                    txn.commit()
+                    txn = db.begin(write=True)
+            txn.commit()
+
+            keys = [u'{}'.format(k).encode('ascii') for k in range(idx + 1)]
```

Comment:
Should be `keys = [u'{:08}'.format(k).encode('ascii') for k in range(self.latest_idx + 1)]`